Add more Procera interfaces

pheinrichs opened the pull request:
I've added more ports related to the procera mibs. Please review and let me know if you think this could be done any better or if you have any suggestions. 

Thanks :)

DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/librenms/librenms/7422)
<!-- Reviewable:end -->

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/6bbf120d-facb-43fb-bdcc-381fd29e3883) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/6bbf120d-facb-43fb-bdcc-381fd29e3883) completed: **No new issues**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/fa960fd8-67c4-4936-b80a-3f8746d5f1ef) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/fa960fd8-67c4-4936-b80a-3f8746d5f1ef) completed: **No new issues**

murrant commented on includes/definitions/procera.yaml:
```diff
@@ -11,3 +11,5 @@ processor_stacked: 1
 discovery:
    - sysObjectId:
        - .1.3.6.1.4.1.15397.2
+mib_dir:
```
MIB dirs matching the OS name are added by default.

pheinrichs replied:
Well that's convenient. I'll remove it, thanks Murrant.

murrant commented on includes/polling/ports/procera.inc.php:
```diff
@@ -0,0 +1,64 @@
+*/
+
+$packetlogic_stats = snmpwalk_cache_oid($device, 'netDeviceTable', array(), 'PACKETLOGIC-CHANNEL-MIB');
+$packetlogic_info = snmpwalk_cache_oid($device, 'channelInfoTable', array(), 'PACKETLOGIC-CHANNEL-MIB');
```
If you put both these in the same array it will make things easier for you below.  Just send the array from the first walk to the second walk instead of array().

snmpwalk_group() is a newer function, but either is fine here.

laf replied:
I think I'd like to see the use of `snmpwalk_group()` if possible.

pheinrichs replied:
I attempted the snmpwalk_group(), I don't know if the oddly decimal labeled indexes are causing issues but here's what happened, also created some odd loop that created 6 more interfaces :confused: 

https://pastebin.com/eADLHYij

Did I use it wrong?

murrant replied:
That is odd @pheinrichs can you do that same thing with debug set so I can see the raw snmpwalk output?

pheinrichs replied:
Adding the depth at 2 seems to be kind of working but still giving the array a sub array with an index of [.0] 

https://pastebin.com/0gSBzhwq

murrant replied:
Yeah, the snmp engine isn't returning it right.

It is returning `channelInternalTxAborted[11].0` when it should be `channelInternalTxAborted[11][0]`

Maybe I can update the code to handle this as it might be common when walking tables of single values.

murrant replied:
#7427 should fix this

murrant commented:
The approach seems reasonable to me.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/655f410b-6487-4771-898b-f05750dbabe2) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/655f410b-6487-4771-898b-f05750dbabe2) completed: **No new issues**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/452af138-d0ea-4b95-8e24-b4364610c8cd) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/452af138-d0ea-4b95-8e24-b4364610c8cd) completed: **No new issues**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/a74ca38c-845c-4d58-b9cc-5237a74174c6) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/a74ca38c-845c-4d58-b9cc-5237a74174c6) completed: **No new issues**

laf commented on includes/polling/ports/procera.inc.php:
```diff
@@ -0,0 +1,64 @@
+        $procera_port['ifConnectorPresent'] = ($packetlogic_stats[$index][0]['NegotiatedMedia'] != 'linkdown' ? "true" : "false");
+        $procera_port['ifOperStatus'] = ($packetlogic_stats[$index][0]['channelActive'] === 'active' ? "up" : "down");
+        $procera_port['ifType'] = 'ethernetCsmacd';
+        array_push($port_stats, $procera_port);
```
This will add it to the existing array, are these ports in addition to? If so, do we need anything like speed and things to be registered?

murrant replied:
channelInternalNegotiatedMedia should allow speed/duplex to be filled in.

But yes, please try to fill in as many fields from ifTable as possible.

murrant commented on includes/polling/ports/procera.inc.php:
```diff
@@ -0,0 +1,64 @@
+* @author     Paul Heinrichs <pdheinrichs@gmail.com>
+*/
+
+$packetlogic_stats = snmpwalk_group($device, 'netDeviceTable', 'PACKETLOGIC-CHANNEL-MIB', 2, array());
```
depth should be 1, so you can omit it and array() in the first call.

murrant commented:
@pheinrichs You might want to merge or rebase this against master to pull in the snmpwalk_group() changes.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/f1627730-6427-466d-92c8-2a51216d699c) was created.

pheinrichs commented:
@murrant @laf  I've added IfSpeed / duplex and also changed the depth to 1 in the snmpgroup_walk. I also revisited the PACKETLOGIC-CHANNEL-MIB to make sure I didn't miss any extra fields. Let me know what you think! :)

murrant commented on includes/polling/ports/procera.inc.php:
```diff
@@ -46,19 +46,33 @@
+        $procera_port['ifName'] = $packetlogic_stats[$index]['channelName']. ' '.$cType['name'];
+        $procera_port['ifDescr'] = $packetlogic_stats[$index]['channelName']. ' '.$cType['name'];
+        $procera_port['ifConnectorPresent'] = ($packetlogic_stats[$index][$cType['type']."NegotiatedMedia"] != '0' ? "true" : "false");
+        $procera_port['ifOperStatus'] = ($packetlogic_stats[$index]['channelActive'] == true ? "up" : "down");
```
`$packetlogic_stats[$index]['channelActive'] == 1` might be a little more clear.  This oid returns 1 or 0.

murrant commented on includes/polling/ports/procera.inc.php:
```diff
@@ -46,19 +46,33 @@
-        $procera_port['ifOperStatus'] = ($packetlogic_stats[$index][0]['channelActive'] === 'active' ? "up" : "down");
+        $procera_port['ifName'] = $packetlogic_stats[$index]['channelName']. ' '.$cType['name'];
+        $procera_port['ifDescr'] = $packetlogic_stats[$index]['channelName']. ' '.$cType['name'];
+        $procera_port['ifConnectorPresent'] = ($packetlogic_stats[$index][$cType['type']."NegotiatedMedia"] != '0' ? "true" : "false");
```
NegotiatedMedia is used 3 times, a temporary variable might be in order.

murrant commented:
@pheinrichs I'm testing this on our PL7810.

The ports from ifMib are mostly useless and duplicates, do you think we should remove them?  They do contain valid MAC info though...

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/d2234ac2-4720-4a06-96f9-f335e3c56987) was created.

pheinrichs commented:
@murrant Them containing macs is what makes me hesitant. I'd vote for leave them?

murrant commented:
I'm saying you should merge them together.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/d2234ac2-4720-4a06-96f9-f335e3c56987) completed: **No new issues**

pheinrichs commented:
@murrant sorry for the delay. I don't think i understand what you would like done. Is there an example i could look off of.

murrant commented:
um, I think the airfiber devices do it.  

basically, instead of presenting 16 ports to the user (when there is physically 8), present 8 ports to the user, taking the correct parts of data from if-mib and the procera mibs and combining them.

Basically what you are doing here, but also pull in the if-mib data.

pheinrichs commented:
I've been comparing the two, and they don't really match up. the IF-MIB generates about 10 ports, but there are only about 3 virtual ports(3 internal/ 3 external) that get discovered from my PR. Is this still the desired way to go?

murrant commented:
I looked at it and you're right it doesn't line up.

I've thought about this and think the best solution is to add this to the yaml file:
```
bad_if_regexp: '/^pl[0-9]+$/'
```

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/d05089eb-3462-4e00-a857-e9a79110bcb8) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/d05089eb-3462-4e00-a857-e9a79110bcb8) completed: **No new issues**
